Feat/collection

wildintellect opened the pull request:
Before you submit a pull request, please fill in the following:

**Related Issue(s):**
#7 

**Description:**
New Feature: Add function to generate Collection record.
Bonus: Allow for the HOST Provider to be selected from options.

**PR checklist:**

- [x] Code is formatted (run `scripts/format`).
- [x] Code lints properly (run `scripts/lint`).
- [x] Tests pass (run `scripts/test`).
- [x] Documentation has been updated to reflect changes, if applicable.
- [x] Changes are added to the [CHANGELOG](../CHANGELOG.md).
- [x] (@gadomski) Update Github settings for new Python version matrix

wildintellect committed:
chore:update gitignore for vscode

wildintellect committed:
chore:clean linting

editorconfig binary is now ec, todo:consider upgrading from stactools template

wildintellect committed:
chore:make yapf on CI work

yapf is looking at pyproject.toml, but toml isn't installed, and yapf section isn't specified

wildintellect committed:
feat:inital collection function

Added constants and basic collection, TODO: add all values, verify values, handle different hosts

wildintellect committed:
fix:cli args correction

Click wants arguments to be in the docstring. Also bumped the version number pre-emptively to avoid confusion, started Changelog

wildintellect committed:
feat: refactor constant import

importing all constants now, in progress - adding all the collection components

wildintellect committed:
feat: add tests

Added stubs for Collection tests

wildintellect committed:
feat:Add license and description

Added new fields, fixed existing fields to pass tests, TODO: additional fields, adjust host

wildintellect committed:
fix:full tests working

Command and Stac tests both work now, key was making the CatalogType relative

wildintellect committed:
feat: add option to specificy host

Host is now a cli argument on collections to allow the same recipe to be used on multiple implementations

wildintellect committed:
feat: Add host selection to Items too

wildintellect committed:
Merge branch 'main' into feat/collection

wildintellect committed:
Merge branch 'feat/collection' into feat/host-changer

wildintellect committed:
typo: Singer/Sinerg

wildintellect committed:
chore: Update Changelog

wildintellect committed:
fix:#1 corrections

Added missing Collection on Items, added links to Collection

wildintellect commented:
Should #6 adding of assets be included in this?
Possible TODO: modify documentation.

wildintellect committed:
feat: Improve cli handling of options

host is optional but when supplied comes from a list, product also is limited to a specific list

wildintellect committed:
chore: Validate Host on items too

wildintellect committed:
Merge pull request #4 from developmentseed/feat/host-changer

Feat/host changer

wildintellect commented:
TODO: 
- [x] bump the minimum version of Python to 3.8, the `:=` operator was introduced then.
- [x] update documentation

wildintellect committed:
fix: remove python 3.7, add 3.10

 := was added in 3.8

wildintellect committed:
chore:update readme/changelog/examples

wildintellect committed:
chore:missing quotes on python-version

wildintellect committed:
chore:linting line endings

wildintellect committed:
chore:deps for python 3.10

wildintellect committed:
feat:Add VS Code container config

wildintellect committed:
typo:devcontainer context fix

wildintellect commented:
@matthewhanson @gadomski not sure who the best person to review this is.

gadomski commented on CHANGELOG.md:
```diff
@@ -4,7 +4,28 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project attempts to match the major and minor versions of [stactools](https://github.com/stac-utils/stactools) and increments the patch number as needed.
 
-## [Unreleased]
+## [0.1.1]
+
+### Added
+
+- create-collection and matching tests for both glo-30 and glo-90
+- New: `--host` option to specify a Provider HOST from a list of options in constants. `AWS` and `OT` are the current valid options.
+- added python 3.10
+
+### Deprecated
+
+- Nothing
+
+### Removed
+
+- remove python 3.7 support
+
+### Fixed
+
+- Nothing.
+
+
+## [0.1.0]
```
Need to add links to the bottom of the CHANGELOG for these tags -- see https://raw.githubusercontent.com/stactools-packages/usda-cdl/main/CHANGELOG.md for an example of the href format.

wildintellect replied:
There's no tag for 0.1.0 right now, we should tag that if we want to be able to generate those diff links.

gadomski replied:
Yeah, it looks like it wasn't ever released (https://pypi.org/project/stactools-cop-dem/ is a 404). In that case, this can all go in the `[Unreleased]` section, and we can make the CHANGELOG pretty before the first release.

gadomski commented on examples/cop-dem-glo-30.json:
```diff
@@ -0,0 +1,97 @@
+  "stac_extensions": [
+    "https://stac-extensions.github.io/item-assets/v1.0.0/schema.json",
+    "https://stac-extensions.github.io/projection/v1.0.0/schema.json",
+    "https://stac-extensions.github.io/raster/v1.1.0/schema.json"
```
I don't see any `raster` properties in this collection -- if the extension is enabled, I'm guessing there's supposed to be `raster:bands` on the item asset?

wildintellect replied:
Now added, minimally.

gadomski commented on examples/cop-dem-glo-30.json:
```diff
@@ -0,0 +1,97 @@
+      "href": "https://spacedata.copernicus.eu/documents/20126/0/CSCDA_ESA_Mission-specific+Annex.pdf",
+      "type": "Copernicus Data Access"
+    },
+    {
+      "rel": "self",
+      "href": "/workspaces/cop-dem/cop-dem-glo-30/cop-dem-glo-30.json",
+      "type": "application/json"
+    }
```
For examples, we usually exclude the `self` href, since it's an absolute path that won't make sense on most systems (e.g. this one's for the dev container, but doesn't make sense when accessed via its raw Github href).

`include_self_href=False` could be enabled via a CLI option, but I've often found it easier to make a standalone script to generate the examples, e.g. https://github.com/stactools-packages/usda-cdl/blob/main/scripts/create_examples.py.

wildintellect replied:
fixed by adding a script that normalized as you suggested

gadomski commented on src/stactools/cop_dem/stac.py:
```diff
@@ -1,30 +1,33 @@
 import re
 from typing import Optional
-from pystac.asset import Asset
+# from pystac.asset import Asset
```
If we don't need it, just remove it.
```suggestion
```

gadomski commented on src/stactools/cop_dem/stac.py:
```diff
@@ -42,18 +45,37 @@ def create_item(href: str,
-    item.add_links(COP_DEM_LINKS)
-    item.common_metadata.platform = COP_DEM_PLATFORM
+    item.add_links(co.COP_DEM_LINKS)
+    # Valid stac item requires collection
+    # collection requires a matching Link object
+    item.collection_id = collection
+    root_href = ''  # TODO: should this become and optional arg?
+    item.links.append(
+        Link(rel="collection",
+             target=os.path.join(root_href, f"{collection}.json")))
```
```suggestion
```

I'd suggest removing this section. Items are usually wired into collections by a higher-level script after item creation -- different platforms might end up having different values for the collection id field, for example.

wildintellect replied:
removed

gadomski commented on src/stactools/cop_dem/__init__.py:
```diff
@@ -8,4 +8,4 @@ def register_plugin(registry):
 
 
-__version__ = "0.1.0"
+__version__ = "0.1.1"
```
```suggestion
__version__ = "0.1.0"
```

Let's not bump the version yet -- IMO it's best to release off of a standalone PR, so it's easier to audit the changed and ensure (e.g.) there is nothing breaking that would require a v0.2.

wildintellect replied:
Sounds good.

gadomski commented on CHANGELOG.md:
```diff
@@ -4,7 +4,28 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project attempts to match the major and minor versions of [stactools](https://github.com/stac-utils/stactools) and increments the patch number as needed.
 
-## [Unreleased]
+## [0.1.1]
```
```suggestion
## [Unreleased]
```

Per my comment on the `__init__.py` file, let's not release quite yet.

gadomski commented:
Once this PR is approved, I'll update the Github actions settings for the new Python version matrix.

wildintellect committed:
fix: some PR review changes

Fixes version number, extra import

wildintellect committed:
feat: Adding Raster ext to item(draft), fixing cli typos

Added the Raster Extension but not convinced 256 bin histogram is useful, and units aren't included. Manually marked PR as draft until remaining issues are addressed

wildintellect committed:
feat:Add RasterBand, remove collection ref

Requested changes to PR adding Raster ext completed, remove collection ref in item

wildintellect committed:
feat: script examples

Based on PR feedback moved example creation to script, removed self references to make portable. Now includes collection and itesm for both glo-30 and glo-90

gadomski commented on examples/cop-dem-glo-30/collection.json:
```diff
@@ -0,0 +1,97 @@
+  "stac_extensions": [
+    "https://stac-extensions.github.io/item-assets/v1.0.0/schema.json",
+    "https://stac-extensions.github.io/projection/v1.0.0/schema.json",
+    "https://stac-extensions.github.io/raster/v1.1.0/schema.json"
```
I still don't see any `raster:` attributes in this collection, so we should either add some (e.g. `raster:bands` on the `item_assets`) or remove this schema.

wildintellect replied:
odd, I added https://github.com/developmentseed/cop-dem/blob/fee76e0797db5ef4a2d2efaa3c350600b23289a6/src/stactools/cop_dem/stac.py#L84-L89 I'll go figure out why that didn't make it into the examples at the end.

wildintellect replied:
fixed, I think there was a silent failure on permissions when I ran the new create_example.py that failed to overwrite the old version.

wildintellect committed:
fix:reran create_example to get missing raster:

philvarner commented:
was #7 resolved by this (it's still open)?

gadomski commented:
This issue is closed:
- https://github.com/stactools-packages/cop-dem/issues/7

There are collections in the examples (e.g. https://github.com/stactools-packages/cop-dem/blob/main/examples/cop-dem-glo-90/collection.json), and there is a `create-collection` command: https://github.com/stactools-packages/cop-dem/blob/f7e2207f610f9c5ca2fa828b9d22e5a106493b9b/src/stactools/cop_dem/commands.py#L42-L75

So I think, yes?
